### PR TITLE
[MoE][6/n] Extract local_reorder, split DeepEP/HybridEP dispatchers

### PR DESCRIPTION
**Pull request by acisseJZhong**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #3389
* __->__ #3386
* #3385

## Summary
MoE cleanup from https://github.com/pytorch/torchtitan/issues/2956:

- **Extract `local_reorder` helper** on `LocalTokenDispatcher`: deduplicates the histc/argsort/score-weighting block between `LocalTokenDispatcher.dispatch()` and `AllToAllTokenDispatcher.dispatch()`
- **Split `DeepEPTokenDispatcher` into `DeepEPTokenDispatcher` + `HybridEPTokenDispatcher`**: eliminates `comm_backend` string branching; config fields (`non_blocking_capacity_factor`, `pad_multiple`) are now class-specific. Callsites use `isinstance` instead of `getattr(..., "comm_backend", ...)`
- **Fix DeepEP CUDA stream race** (#3413): added `torch.cuda.synchronize()` before `get_dispatch_layout` in `dispatch_tokens()` to ensure `masked_fill` results are visible on DeepEP's comm stream. Without this, `get_dispatch_layout` reads stale (pre-masked) expert indices, causing `num_tokens_per_expert` to overcount and `_grouped_mm` to assert.

## Test plan
Test loss comparions 
```
NCCL_NVLS_ENABLE=0 python scripts/loss_compare.py . main --baseline-module='llama4' --baseline-config='llama4_debugmodel_ep' --baseline-options="--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2 --parallelism.no-enable_sequence_parallel" --baseline-ngpus=4 --test-module='llama4' --test-config='llama4_debugmodel_ep' --test-options="--parallelism.pipeline_parallel_degree 1 --parallelism.data_parallel_shard_degree 2 --parallelism.tensor_parallel_degree 2 --parallelism.expert_parallel_degree 2 --parallelism.no-enable_sequence_parallel" --test-ngpus=4 --assert-equal --no-seed-checkpoint
```

### CODE REVIEW

File: torchtitan/models/deepseek_v3/model.py [hunk 1]
```diff
@@ -214,17 +218,13 @@ def update_from_config(
                     layer_cfg.moe.router._debug_force_load_balance = (
                         debug.moe_force_load_balance
                     )
-                    comm_backend = getattr(
-                        layer_cfg.moe.experts.token_dispatcher,
-                        "comm_backend",
-                        "standard",
-                    )
+                    td_cfg = layer_cfg.moe.experts.token_dispatcher
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model.py]:
```suggestion
                    token_dispatcher_cfg = layer_cfg.moe.experts.token_dispatcher
```

File: torchtitan/models/llama4/model.py [hunk 2]
```diff
@@ -139,17 +143,13 @@ def update_from_config(
                     layer_cfg.moe.router._debug_force_load_balance = (
                         debug.moe_force_load_balance
                     )
-                    comm_backend = getattr(
-                        layer_cfg.moe.experts.token_dispatcher,
-                        "comm_backend",
-                        "standard",
-                    )
+                    td_cfg = layer_cfg.moe.experts.token_dispatcher
```
Comment (tianyu-l) [inline torchtitan/models/llama4/model.py]: similar

File: torchtitan/models/common/token_dispatcher.py [hunk 3]
```diff
@@ -564,24 +624,32 @@ class Config(LocalTokenDispatcher.Config):
...
 
         # Import to register custom ops so SAC saves communication outputs
         # instead of recomputing them. This must happen before apply_ac.
-        if config.comm_backend == "hybridep":
-            from torchtitan.distributed.deepep import hybridep  # noqa: F401
-        else:
-            from torchtitan.distributed.deepep import deepep  # noqa: F401
+        from torchtitan.distributed.deepep import hybridep  # noqa: F401
+
+    def wire_meshes(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:689]: did you mean to put this into DeepEP dispatcher?

File: torchtitan/models/common/token_dispatcher.py [hunk 4]
```diff
@@ -64,27 +65,29 @@ def wire_meshes(
         """No-op for the EP=1 dispatcher. Subclasses override."""
         del ep_mesh, tp_mesh
 
-    def dispatch(
+    def local_reorder(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]:
```suggestion
    def _local_reorder(
```

File: torchtitan/models/common/token_dispatcher.py
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:447]: To double check
- The Expert SP initial sharding is handled uniformly by config now, in the last PR.
- Do we not need this in DeepEP / HybridEP token combine, because it's handled correctly internally?

Comment (acisseJZhong) [reply torchtitan/models/common/token_dispatcher.py:447]: > The Expert SP initial sharding is handled uniformly by config now, in the last PR.

Yes this correct. I removed the manual SP split, sharding should be handled by config now.
```
    moe_desired_input_layouts = Replicate() if not enable_ep else sp_layout
```

> Do we not need this in DeepEP / HybridEP token combine, because it's handled correctly internally?

I assume it's handled internally by deepep/hybridep? But now that you brining up this I am not sure if it could handle SP? I need to look deeper. 

Previously we were not manually splitting for deepep / hybridep, but now config will shard input so I am not sure how deepep / hybrid handles it.

File: torchtitan/distributed/utils.py [hunk 5]
```diff
@@ -132,6 +132,8 @@ def set_determinism(
         )
         torch.backends.cudnn.deterministic = True
         torch.backends.cudnn.benchmark = False
+        # DeepEP's comm streams conflict with torch.empty() init kernels
+        torch.utils.deterministic.fill_uninitialized_memory = False
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py:140]: can we remove https://github.com/pytorch/torchtitan/blob/963c20cba37f392fcfdff3b1c7519fde8ad4c0a7/torchtitan/experiments/transformers_modeling_backend/model.py#L289-L292

File: torchtitan/distributed/utils.py [hunk 6]
```diff
@@ -132,6 +132,10 @@ def set_determinism(
         )
         torch.backends.cudnn.deterministic = True
         torch.backends.cudnn.benchmark = False
+        # torch.empty() init kernels overlap with DeepEP comm streams and
+        # initialize HF ROPE (inv_freq) buffers to NaN when this is True.
```
Comment (tianyu-l) [inline torchtitan/distributed/utils.py]: separate the comments: this is only for HF models; the line above is only for DeepEP

**Review by tianyu-l — APPROVED** [review]
